log4j2/src/main/java/com/google/common/flogger/backend/log4j2/ValueList.java

```diff
@@ -0,0 +1,156 @@
+                    : new ValueList(existingValue, null);
+            return new ValueList(value, existingList);
+        }
+    }
+
+    @Override
+    public Iterator<Object> iterator() {
+        return new LinkedIterator(this);
+    }
+
+    static void appendValues(String label, Object valueOrList, MetadataKey.KeyValueHandler kvh) {
+        if (valueOrList instanceof ValueList) {
+            for (Object value : (ValueList) valueOrList) {
+                emit(label, value, kvh);
+            }
+        } else {
+            emit(label, valueOrList, kvh);
+        }
+    }
+
+    /**
+     * Tags are key-value mappings which cannot be modified or replaced. If you add
+     * the tag mapping "foo" -> true and later add the mapping "foo" -> false, you
+     * get the value "foo" mapped to both true and false. This is *very* deliberate
+     * since the key space for tags is global and the risk of two bits of code
+     * accidentally using the same tag name is real (e.g. you get "id=abcd" but you
+     * added "id=xyz" so you think this isn't your log entry, but someone else added
+     * "id=abcd" in a context you weren't aware of).
+     *
+     * Given three tag mappings: "baz" -> , "foo" -> true and "foo" -> false the
+     * Value list is going to flatten the tags in order to store them as tags=[baz,
+     * foo=false, foo=true].
+     * 
+     * Reusing the label 'tags' is intentional as this allows us to store the
+     * flatten tags in Log4j2's ContextMap.
+     */
+    static void emit(String label, Object value, MetadataKey.KeyValueHandler kvh) {
+        if (value instanceof Tags) {
+            // Flatten tags to treat them as keys or key/value pairs, e.g. tags=[foo,
+            // baz=bar2, baz=bar]
+            ((Tags) value).asMap().forEach((k, v) -> {
+                if (v.isEmpty()) {
+                    kvh.handle(label, k);
+                } else {
+                    for (Object obj : v) {
+                        kvh.handle(label, k + "=" + obj.toString());
+                    }
+                }
+            });
+        } else {
+            kvh.handle(label, value);
+        }
+    }
+
+    /**
+     * Returns a string representation of the contents of the specified value list.
+     * If the list contains other lists as elements, they are converted to strings.
+     *
+     * <p>
+     * If value list is empty, this method returns {@code "null"}.
+     * <p>
+     * If value list contains a single element {@code a}, this method returns
+     * {@code "a"}.
+     * <p>
+     * Otherwise the content of teh list is returned surrounded by brackets "[", "].
+     */
+    @Override
+    public String toString() {
+        if (value == null) {
+            return "null";
+        }
+
+        if (next == null || next.value == null) {
+            return value.toString();
+        }
+
+        StringBuilder out = new StringBuilder();
```

Comment:
If toString() is rare and doesn't need to be efficient, this could be:

return StreamSupport.stream(spliterator()).collect(Joiner.joining(", ", "[", "]"); 

And I would never return "null" or anything like that. Once we have a list, it should always look like `[ ... ]` even if one element or empty (though empty shouldn't actually happen I think), you don't really need a comment on the `toString()` method because it's just like any other list in Java.

Reply:
I checked it out, but I would like to propose a slightly different solution.
The current solutions treats Tags and Metadata slightly differently:
- Tags will always be formatted as a list of values
- Depending on the content of the metadata, metadata will be usually formatted as a single key/value-pair or a list of values.

As far as I can tell the `maybeWrap` proposol covers the handling of metadata well. Following this train of thought, I find it a bit more intuitive to print 
- "" (empty string) when the queue is empty, 
- a single value, when the queue contains a single value only, and
- a list of values when the queue contains n > 1 elements

instead of an empty list when the queue is empty (though the Tags are treated differently by explicitly wrapping a single tag into a list).